# Exercise 5

Related Notes:
- Chapter_05_File_IO

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/njc-cz2-2021/Materials/blob/main/Exercises/Exercise_05_File_IO.ipynb) 

## Exercise 5.1

Write following list `s` into a text file `q1.txt` with each item occupying 1 line.

>``` python
>s = [chr(i) + '.txt' for i in range(ord('A'), ord('F'))]
>```

Use following command in Jupyter Notebook to open `q1.txt` to verify the content in the file.

>`!notepad q1.txt`

In [1]:
#YOUR_CODE_HERE
!notepad q1.txt
s = [chr(i) + '.txt' for i in range(ord('A'), ord('F'))]
with open('q1.txt','a') as f:
    for i in range(len(s)):
        f.write(f'{s[i]}\n')

## Exercise 5.2

For each line of text in `q1.txt`, generate an empty text file and name the file using that line of text. If file already exists, remove it before generating a new file. 

Use the following code to check if a file exists.

>``` python
>import os
>if os.path.exists(file_path): pass
>```

Use following command in Jupyter Notebook to confirm that you have generated all files.

>``` bash
>!ls *.txt
>```

In [ ]:
#YOUR_CODE_HERE
import os 
with open('q1.txt',r):
    

## Exercise 5.3

Write code to generate all the positive integers up to $n$ that are perfect squares and save them in a text file called `PERFECT_SQUARES.TXT`. Note that each line within the file should contain one integer, and the last character within the file should be an end line character.

In [ ]:
#YOUR_CODE_HERE

## Exercise 5.4

Write code to generate all the positive integers from $n + 1$ to $n + m$, and then append them to the file used in question 1. (Note, do not overwrite the perfect squares from Question 1.) 

In [ ]:
#YOUR_CODE_HERE

## Exercise 5.5

A log of a certain boss fight from a certain game is stored inside the file `log_ex8.csv` under the folder `resources`. The csv file contains the following header.

- `'Parse%'`  refers to the performance of the character relative to all other players in the world,
- `'Name'` refers to the name of the character in the game,
- `'Amount'`  refers to the amount of damage
- `'iLvL'`  refers to the quality of the equipment the character is using,
- `'iLvL%'` refers to the performance of the player relative to players with similar quality of equipment,
- `'Active Time'` refers to the percentage of time spend by the character doing damage during the fight.
- `'DPS'` refers to the amount of damage done per second in the fight.

Your task is to:

- Store the contents of the csv file into a list `records` where each element in the list is itself a list where the elements correspond to the `Parse%`,`Name`,`Amount`,`iLvL`,`iLvL%`,`Active Time`,`DPS` values for each line in the csv file.
- Create an empty list called `player_name`.
- iterate over each element of `records` appending the player names to `player_name`.
- Print each name in `player_name` in alphabetical order.

In [ ]:
#YOUR_CODE_HERE
with open('resources/log_ex8.csv','r') as f:
    list = f.read()
    list = list.split('\n')
print(list)
empty_list = []
for i in list:
    empty_list.append(i.split(','))
print(empty_list)
player_name = []
for i in range(1,len(empty_list)):
    player_name.append(empty_list[i][1])
player_name=sorted(player_name)
  
    




['"Parse %","Name","Amount","Ilvl","Ilvl %","Active","DPS",""', '"97","Wôes","11715922$6.32%11.72m","208","82","95.30%","4,857.4",""', '"94","Dewgie","10230823$5.52%10.23m","-","57","92.34%","4,241.7",""', '"96","Knoxvegas","9624618$5.19%9.62m","206","77","94.72%","3,990.3",""', '"92","Avey","9384888$5.06%9.38m","208","54","95.44%","3,890.9",""', '"77","Exolaris","8817948$4.76%8.82m","198","75","96.15%","3,655.9",""', '"79","Ovy","8789252$4.74%8.79m","-","89","91.45%","3,644.0",""', '"77","Biraprie","8684125$4.69%8.68m","-","75","94.53%","3,600.4",""', '"88","Kolhito","8320548$4.49%8.32m","-","76","90.10%","3,449.7",""', '"67","Fyrebird","8286557$4.47%8.29m","-","80","96.52%","3,435.6",""', '"67","Jari","8179735$4.41%8.18m","-","69","94.69%","3,391.3",""', '"69","Anothervc","8178271$4.41%8.18m","-","88","91.36%","3,390.7",""', '"73","Stormblast","7797837$4.21%7.80m","191","83","88.15%","3,232.9",""', '"64","Spooknspell","7772714$4.19%7.77m","189","91","93.02%","3,222.5",""', '"59","Pl

## Exercise 5.6

Download the file: https://archive.ics.uci.edu/ml/machine-learning-databases/iris/iris.data

Write code to read and store the contents of this comma separated file within a 2D List, $L$, such that $L[i]$ corresponds to the sample entry on the $i$-th line of the file, and $L[i][j]$ corresponds to the $j$-th value of the $i$-th entry.
Print the contents of the 2D List to verify that you have stored the contents correctly.


In [ ]:
#YOUR_CODE_HERE
with open('resources/IRIS.data','r') as f:
    s = f.read()
    s = s.split('\n')
L = []
for i in s:
    L.append(i.split(','))
print(L)

        
    

[['5.1', '3.5', '1.4', '0.2', 'Iris-setosa'], ['4.9', '3.0', '1.4', '0.2', 'Iris-setosa'], ['4.7', '3.2', '1.3', '0.2', 'Iris-setosa'], ['4.6', '3.1', '1.5', '0.2', 'Iris-setosa'], ['5.0', '3.6', '1.4', '0.2', 'Iris-setosa'], ['5.4', '3.9', '1.7', '0.4', 'Iris-setosa'], ['4.6', '3.4', '1.4', '0.3', 'Iris-setosa'], ['5.0', '3.4', '1.5', '0.2', 'Iris-setosa'], ['4.4', '2.9', '1.4', '0.2', 'Iris-setosa'], ['4.9', '3.1', '1.5', '0.1', 'Iris-setosa'], ['5.4', '3.7', '1.5', '0.2', 'Iris-setosa'], ['4.8', '3.4', '1.6', '0.2', 'Iris-setosa'], ['4.8', '3.0', '1.4', '0.1', 'Iris-setosa'], ['4.3', '3.0', '1.1', '0.1', 'Iris-setosa'], ['5.8', '4.0', '1.2', '0.2', 'Iris-setosa'], ['5.7', '4.4', '1.5', '0.4', 'Iris-setosa'], ['5.4', '3.9', '1.3', '0.4', 'Iris-setosa'], ['5.1', '3.5', '1.4', '0.3', 'Iris-setosa'], ['5.7', '3.8', '1.7', '0.3', 'Iris-setosa'], ['5.1', '3.8', '1.5', '0.3', 'Iris-setosa'], ['5.4', '3.4', '1.7', '0.2', 'Iris-setosa'], ['5.1', '3.7', '1.5', '0.4', 'Iris-setosa'], ['4.6', '

## Exercise 5.7

Each line in the file at: http://itu.dk/people/pagh/sad12/primes1000000.txt represents a prime number below 1,000,000.
 
Write a function that uses the above file to test if a given positive integer below 1,000,000 is a prime number. If the given numbers exceeds 1,000,000, then utilise a separate function to calculate if it is a prime. 

In [24]:
#YOUR_CODE_HERE
number = int(input('input an integer')) 
import math
def check_prime(n):
 if n<= 1:
    print(False)
 else:
    is_prime = True
    for i in range(2, int(math.sqrt(n)) + 1):
        if n % i == 0:
            is_prime = False
            break
    print(is_prime)

def check_prime_file(n):
    n = str(n)
    with open('resources/primes1000000.txt','r') as f:
        s = f.read()
        s = s.split('\n')
        
    if n in s:
            print(True)
    else:
            print(False)
 
if number <1000000:
    check_prime_file(number)
else:
    check_prime(number)

False


## Exercise 5.8

A mathematical equation $y=f(x)$ can be visualized on a cartesian plane as the set of all the points $(x,y)$ where $y=f(x)$. Write a code that 

1. generates 100 points that satisfy the equation $y=x^2$ for $-10\leq x\leq 10$ (for simplicity, take the uniform spacing for the $x$-values of the points),and
2. saves them into a text file named `ex8_8.txt`. Each line in the file is of the form $(x,y)$, 
3. load the points from `ex8_8.txt` and using the module `matplotlib`, show the plot.


In [ ]:
#YOUR_CODE_HERE
import matplotlib
x_count = -10
with open('ex8_8.txt','a') as f:
    for i in range(100):
        f.write(f'{x_count+(20/100*i)},{(x_count+(20/100*i))**2}')


    

## Exercise 5.9

A text file, `NOI_24_RESULT.txt` contains the results of the medallist recipients for the 24th National Olympiad in Informatics.

Write a code to:

- read the data from the text file,
- store the school name and medal counts in a Python dictionary `medal_count` where the school name is the key and the associated value is a tuple of the form `(GOLD, SILVER, BRONZE)`,
- print out the top 5 school with the highest number of medals won, regardless of the medal type. 

In [ ]:
#YOUR_CODE_HERE
#jovan one very funni
import csv

Gold = []
Silver = []
Bronze = []

amount_of_medals_school = []

with open("./resources/NOI_24_RESULT.csv") as f:
    s = csv.reader(f)
    print(s)
    for i in s:
        for n in i:
         n.replace('\n','')
    list_of_medals = [row for row in s if row != []]
    
    gold_start = list_of_medals.index(['Gold']) + 1
    silver_start = list_of_medals.index(['Silver'])
    bronze_start = list_of_medals.index(['Bronze'])

    Gold = [row[0] for row in list_of_medals[gold_start:silver_start]]
    Silver = [row[0] for row in list_of_medals[silver_start + 1:bronze_start]]
    Bronze = [row[0] for row in list_of_medals[bronze_start + 1:]]

    for i in Gold:
        _, school = i.split(',', 1)
        amount_of_medals_school.append([item.strip() for item in school.split(',')][-1])
    
    for i in Silver:
        _, school = i.split(',', 1)
        amount_of_medals_school.append([item.strip() for item in school.split(',')][-1])
        
    for i in Bronze:
        _, school = i.split(',', 1)
        amount_of_medals_school.append([item.strip() for item in school.split(',')][-1])
        
    medal_count = {i : amount_of_medals_school.count(i) for i in amount_of_medals_school}
    
    print("Top 5 schools with most medals:")
    for i in range(len(medal_count)):
        s = max(medal_count, key=medal_count.get)
        if s == 'Singapore':
            s = 'School Of Science And Technology, Singapore'
        s_amt = max(medal_count.values())
        print(f"{s}, with {s_amt} medals")
        s = max(medal_count, key=medal_count.get)
        medal_count.pop(s)

ValueError: ['Gold'] is not in list

In [8]:
Gold = []
Silver = []
Bronze = []

amount_of_medals_school = []

with open("./resources/NOI_24_RESULT.csv", "r") as f:
    lines = [line.strip() for line in f if line.strip()]  # Read lines, strip spaces, skip empty

    gold_start = lines.index('Gold') + 1
    silver_start = lines.index('Silver')
    bronze_start = lines.index('Bronze')

    Gold = lines[gold_start:silver_start]
    Silver = lines[silver_start + 1:bronze_start]
    Bronze = lines[bronze_start + 1:]

    for medal_list in [Gold, Silver, Bronze]:
        for entry in medal_list:
            try:
                _, school = entry.split(',', 1)
                school_name = [item.strip() for item in school.split(',')][-1]
                amount_of_medals_school.append(school_name)
            except ValueError:
                pass  # Handle bad format gracefully

    medal_count = {school: amount_of_medals_school.count(school) for school in set(amount_of_medals_school)}

    print("Top 5 schools with most medals:")
    for _ in range(min(5, len(medal_count))):
        top_school = max(medal_count, key=medal_count.get)
        medal_amount = medal_count[top_school]
        if top_school == 'Singapore':
            top_school = 'School Of Science And Technology, Singapore'
        print(f"{top_school}, with {medal_amount} medals")
        medal_count.pop(top_school)

Top 5 schools with most medals:
Hwa Chong Institution, with 10 medals
NUS High School Of Mathematics And Science, with 10 medals
Raffles Institution, with 10 medals
United World College South East Asia Dover Campus, with 8 medals
Anglo-Chinese School Independent, with 5 medals


## Exercise 5.10

A text file, `MARKET_PRICES.TXT`, contains the in-game prices for 6 items, Widowbloom, Rising Glory, Marrowroot, Vigil's Torch, Nighshade and Spectral Flask of Power from 1 December 2020 to 3 March 2021. Each line contains tab-delimited data that shows the date and the prices (in copper coins) of the 6 items.

Each line is in the format:

>``` text
>YYYY-MM-DD\tWIDOWBLOOM\tRISING_GLORY\t\MARROWROOT\tVIGILS_TORCH\tNIGHTSHADE\tSPECTRAL_FLASK_OF_POWER\n
>```

- The date is of the form `YYYY-MM-DD`, for example, `2020-12-03` is 3rd December, 2020.
- WIDOWBLOOM, RISING_GLORY, MARROWROOT, VIGILS_TORCH, NIGHTSHADE, SPECTRAL_FLASK_OF_POWER are all integers.
- `\t` represents the tab character
- `\n` represents the newline character

The text file is stored in ascending order of date and time.

### Task 1 

Write a program to:

- read the price data from the text file,
- for each of the 6 items:
    - find the highest price and print this value,
    - find the lowest price and print this value.

### Task 2

Suppose you want to play the market by buying and selling the 6 items above, one way to decide on items to invest in is by using the mean and variance of the daily rate of returns of the items. The latter is used to quantify the risk of holding such item.

For an item $j$, 
- the **daily rate of return** of day $i$ to $i+1$, denoted as $r_i$ is defined to be $$r_i=\frac{p_{i+1}-p_{i}}{p_i},$$ where $p_{i+1}$ and $p_i$ are prices of the item on day $i+1$ and day $i$ respectively,
- the mean of the daily rate of returns, denoted as $\overline{r}$, is just $\frac{\sum r_i}{n}$, where $n$ is the number of daily rate of returns recorded,
- the variance of the daily rate of returns is $$\frac{\sum{(r_i-\overline{r})^2}}{n},$$ where $\overline{r}$ is the mean of the daily rate of returns and again $n$ is the number of daily returns recorded.

Write a program to:

- for each of the 6 items:
    - find the mean of the daily rate of return,
    - find the variance of the daily rate of return.

In [18]:
#YOUR_CODE_HERE
#task 1
with open('resources/MARKET_PRICES.txt','r') as f:
    mp = f.read()
    mp = mp.split()
    for i in range(7):
        mp.remove(mp[0])
listofitemnames=['widowbloom','risingglory','marrowroot','vigilstorch','nightshade','flask']
listofitemvalues=[mp[1::7],mp[2::7],mp[3::7],mp[4::7],mp[5::7],mp[6::7]]
print(f"""
      item:          min:   max:      
      widowbloom     {min(listofitemvalues[0])}  {max(listofitemvalues[0])}
      risingglory    {min(listofitemvalues[1])}   {max(listofitemvalues[1])}
      marrowroot     {min(listofitemvalues[2])}   {max(listofitemvalues[2])}
      vigilstorch    {min(listofitemvalues[3])}  {max(listofitemvalues[3])}
      nightshade     {min(listofitemvalues[4])}  {max(listofitemvalues[4])}
      flask          {min(listofitemvalues[5])} {max(listofitemvalues[5])}""")
def ave_rate_return(i):
    count = 0
    for s in range (len(i)-1):
        count+= (int(i[s+1])-int(i[s]))/int(i[s])
        count = count/(len(i)-1)
    return count
def var_rate_return(n):
    ave_r = ave_rate_return(n)
    count = 0
    for i in range(len(n)-1):
        count+= ((int(n[i+1])-int(n[i]))/int(n[i])-ave_r)**2
    count = count/(len(n)-1)
    return(count)

for i in range(6):
    print(f"""
      mean of rate of return for {listofitemnames[i]} is:
      {ave_rate_return(listofitemvalues[i])}
      variable of rate of return for {listofitemnames[i]} is:
      {var_rate_return(listofitemvalues[i])}
              """)



      item:          min:   max:      
      widowbloom     10003  9965
      risingglory    2000   7679
      marrowroot     2402   8677
      vigilstorch    10086  9325
      nightshade     10310  9857
      flask          100644 98674

      mean of rate of return for widowbloom is:
      0.0018833026890536454
      variable of rate of return for widowbloom is:
      0.022845548652536282
              

      mean of rate of return for risingglory is:
      -0.0006031488023453025
      variable of rate of return for risingglory is:
      0.036505209127751466
              

      mean of rate of return for marrowroot is:
      0.0006764847439680622
      variable of rate of return for marrowroot is:
      0.03564672561543559
              

      mean of rate of return for vigilstorch is:
      -0.0006561981210950313
      variable of rate of return for vigilstorch is:
      0.02777603719357029
              

      mean of rate of return for nightshade is:
      -0.000732954804071